app/etl/Program.cs

```diff
@@ -24,31 +24,72 @@
 Stopwatch timer = new Stopwatch();
 timer.Start();
 var transaction = db.Database.BeginTransaction();
-foreach (var record in records)
+
+var currentRecord = records.GetEnumerator();
+var enumeratorStatus = currentRecord.MoveNext();
+
+while (enumeratorStatus)
 {
     FormattableString query;
 
+    var record = currentRecord.Current;
     var temperature = decimal.Parse(record.recordValue) / 10m;
     var station = stations[record.stationId];
     var location = new Point(station.lon, station.lat) { SRID = 4326 };
 
-    if (record.recordType.Equals("TMAX", StringComparison.Ordinal))
+    var oldRecordType = record.recordType;
+    enumeratorStatus = currentRecord.MoveNext();
+    var nextRecord = currentRecord.Current;
+
+    // Take advantage of the reality that the data is sorted when we get it, so we can merge
+    // data from adjacent lines into a single insert if a station provides both TMIN and TMAX
+    if (enumeratorStatus && nextRecord.stationId == record.stationId && nextRecord.date == record.date && !nextRecord.recordType.Equals(oldRecordType))
     {
+        decimal maxTemperature;
+        decimal minTemperature;
+        // Insert both
+        if (record.recordType.Equals("TMAX", StringComparison.Ordinal) && nextRecord.recordType.Equals("TMIN"))
+        {
+            maxTemperature = temperature;
+            minTemperature = decimal.Parse(nextRecord.recordValue) / 10m;
+        }
+        else if (record.recordType.Equals("TMIN", StringComparison.Ordinal) && nextRecord.recordType.Equals("TMAX"))
+        {
+            minTemperature = temperature;
+            maxTemperature = decimal.Parse(nextRecord.recordValue) / 10m;
+        }
+        else
+        {
+            continue;
+        }
         query = $@"
+        INSERT INTO StationData (StationId, StationName, SusStation, RecordDate, MaxTemperature, MinTemperature, Location)
+        VALUES ({record.stationId}, {station.name}, {station.isSus}, {record.date}, {maxTemperature}, {minTemperature}, {location})
+        ";
+        enumeratorStatus = currentRecord.MoveNext();
+
+    }
+    else

```

Comment:
Will this end up skipping a record if we go into the `else`?

Reply:
I don't think so. The iterator is setup to be ready for the next iteration before the if statement. In the else case, `nextRecord` becomes `currentRecord` on the following loop. If the statement is true, we end up advancing an extra time